### PR TITLE
HIP-584: Copy Freeze/Unfreeze Logics and Precompiles

### PR DESCRIPTION
**Pull request by bilyana-gospodinova**
**Description**:
Changes in this PR:

- Copied FreezeLogic and UnfreezeLogic from hedera-services (there are a couple of differences with the original described below):
  - Used tokenRelationship.setFrozen instead of tokenRelationship.changeFrozenState (like in services)
  - Used store.updateTokenRelationship(tokenRelationship) instead of tokenStore.commitTokenRelationships(List.of(tokenRelationship)) (like in services)
- Created FreezeLogicTest and UnfreezeLogicTest and added unit tests
- Added new beans for FreezeLogic, UnfreezeLogic, FreezeTokenPrecompile, UnfreezeTokenPrecompile in Services Configuration
- Copied AbstractFreezeUnfreezePrecompile from hedera-services (there are a couple of differences with the original described below):
  - Removed class fields and adapted constructors in order to achieve stateless behaviour
  - Run method accepts Store argument in order to achieve stateless behaviour and returns RunResult. It calls two abstract methods (validateSyntax and executeFreezeUnfreezeLogic) to delegate to the corresponding freeze/unfreeze implementation instead of executeForFreeze and executeForUnfreeze methods.
- Copied FreezeTokenPrecompile, UnfreezeTokenPrecompile from hedera-services (there are a couple of differences with the original described below):
  - Removed class fields and adapted constructors in order to achieve stateless behaviour
  - Implements validateSyntax and executeFreezeUnfreezeLogic from AbstractFreezeUnfreezePrecompile
- Created FreezeTokenPrecompileTest, UnfreezeTokenPrecompileTest and added unit tests
- Added integration FREEZE_TOKEN, UNFREEZE_TOKEN test in ContractCallServicePrecompileTest
- Removed ContractCallServiceTest#precompileCallRevertsForEstimateGas because the FreezeTokenPrecompile is now implemented and the test was irrelevant
- ModificationPrecompileTestContract is updated to include unfreezeTokenExternal function.

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-mirror-node/issues/6237
Fixes https://github.com/hashgraph/hedera-mirror-node/issues/5658

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/FreezeTokenPrecompile.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    private static final Function FREEZE_TOKEN_FUNCTION = new Function("freezeToken(address,address)", INT);
+    private static final Bytes FREEZE_TOKEN_FUNCTION_SELECTOR = Bytes.wrap(FREEZE_TOKEN_FUNCTION.selector());
+    private static final ABIType<Tuple> FREEZE_TOKEN_ACCOUNT_DECODER = TypeFactory.create(ADDRESS_PAIR_RAW_TYPE);
+    private final FreezeLogic freezeLogic;
+
+    public FreezeTokenPrecompile(
+            final PrecompilePricingUtils pricingUtils,
+            final SyntheticTxnFactory syntheticTxnFactory,
+            final FreezeLogic freezeLogic,
+            boolean isFreeze) {
```
Comment (0xivanov) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/FreezeTokenPrecompile.java:64]: `isFreeze` should not be set as a field. You can modify the `run` method of the abstract class to check whether the `body` method created `TokenUnfreezeAccountTransactionBody` (isFreeze= false)or `TokenFreezeAccountTransactionBody` (isFreeze=true)

Comment (0xivanov) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/FreezeTokenPrecompile.java:64]: see #6446 as an example

Comment (bilyana-gospodinova) [reply hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/FreezeTokenPrecompile.java:64]: Actually, we don't need this flag at all. I will remove it. Thanks for pointing this out.

**Review by 0xivanov — DISMISSED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractFreezeUnfreezePrecompile.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        /* --- Execute the transaction and capture its results --- */
+        final var store = ((HederaEvmStackedWorldStateUpdater) frame.getWorldUpdater()).getStore();
+        executeFreezeUnfreezeLogic(transactionBody, store);
+
+        return new EmptyRunResult();
+    }
+
+    public abstract ResponseCodeEnum validateSyntax(final TransactionBody transactionBody);
+
+    public abstract void executeFreezeUnfreezeLogic(final TransactionBody transactionBody, final Store store);
```
Comment (0xivanov) [inline hedera-mirror-web3/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractFreezeUnfreezePrecompile.java:67]: I like the abstraction :)

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by 0xivanov — APPROVED** [review]

**Pull request merged**